include/nbl/builtin/hlsl/math/morton.hlsl

```diff
@@ -0,0 +1,423 @@
+        // Then instead of unpacking all the bits for that coordinate as the lowest bits, we unpack them as the highest ones
+        // by shifting everything `BitWidth - bits(coord)` bits to the left, then at the end do a final *arithmetic* bitshift right by the same amount.
+        
+        const U bitsCoord = BitWidth / U(D) + ((coord < BitWidth % D) ? U(1) : U(0)); // <- this computes the ceil
+        U coordVal = U(0);
+        // Control can be simplified by running a bound on just coordBit based on `BitWidth` and `coord`, but I feel this is clearer
+        [[unroll]]
+        for (U valBitIdx = U(coord), coordBit = U(1) << U(coord), shift = U(coord); valBitIdx < BitWidth; valBitIdx += U(D), coordBit <<= U(D), shift += U(D) - 1)
+        {
+            coordVal |= (value & coordBit) << (BitWidth - bitsCoord - shift);
+        }
+        return bit_cast<I, U>(coordVal) >> (BitWidth - bitsCoord);
+    }
+
+    /**
+    * @brief Returns an element of type U by extracting only the highest bit (of the bits used to encode `coord`)
+    *
+    * @param [in] coord The coordinate whose highest bit we want to extract.
+    */
+    NBL_CONSTEXPR_INLINE_FUNC U extractHighestBit(uint16_t coord) NBL_CONST_MEMBER_FUNC
+    {
+        // Like above, if the number encoded in `coord` gets `bits(coord) = ceil((BitWidth - coord)/D)` bits for representation, then the highest index of these
+        // bits is `bits(coord) - 1`
+        const U coordHighestBitIdx = BitWidth / U(D) - ((U(coord) < BitWidth % U(D)) ? U(0) : U(1));
+        // This is the index of that bit as an index in the encoded value
+        const U shift = coordHighestBitIdx * U(D) + U(coord);
+        return value & (U(1) << shift);
+    }
+
+    // ------------------------------------------------------- BITWISE OPERATORS -------------------------------------------------
+    
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator&(NBL_CONST_REF_ARG(this_t) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        this_t retVal = { value & rhs.value };
+        return retVal;
+    }
+
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator|(NBL_CONST_REF_ARG(this_t) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        this_t retVal = { value | rhs.value };
+        return retVal;
+    }
+
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator^(NBL_CONST_REF_ARG(this_t) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        this_t retVal = { value ^ rhs.value };
+        return retVal;
+    }
+
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator~() NBL_CONST_MEMBER_FUNC
+    {
+        this_t retVal = { ~value };
+        return retVal;
+    }
+
+    // Only valid in CPP
+    #ifndef __HLSL_VERSION
+
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator<<(uint16_t bits) NBL_CONST_MEMBER_FUNC
+    {
+        this_t retVal = { value << U(bits) };
+        return retVal;
+    }
+
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator>>(uint16_t bits) NBL_CONST_MEMBER_FUNC
+    {
+        this_t retVal = { value >> U(bits) };
+        return retVal;
+    }
+
+    #endif
```

Comment:
we need `functional.hlsl`-style wrappers for >> and << then their specializations in this file